sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2435,6 +2435,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val REPARTITION_BEFORE_INSERT =
+    buildConf("spark.sql.execution.repartitionBeforeInsert")
+      .internal()
+      .doc("When perform a insert into partitioned table. Turn on this config to " +
+        "insert a repartition by dynamic partition columns to ease pressure on the NameNode " +
```

Comment:
Do you mind how it reduces the pressure on the NMs? I don't think people can understand what happens by reading docs.

I would describe it concisely and easy to understand (e.g., p. 44 - 48 https://www.slideshare.net/TejasPatil1/hive-bucketing-in-apache-spark). It's bucketing but we could resue it for partitioning for such as the number of files and diagrams.

Probably we should describe pros and cons, and when it should be turned on. Feel free to write something in https://github.com/apache/spark/blob/master/docs/sql-performance-tuning.md too.